Добавлен новый элемент "Бизнес-правило"

zkvan opened the pull request:
Добавлен новый элемент "Бизнес-правило". Был изменен модификатор доступа FormulaCellEditorProvider.ConfigurationDialog на public, чтобы избежать дублирования кода.

Commit:
Добавлен новый элемент "Бизнес-правило"

AlexMisha commented on plugins/ru.runa.gpd/schema/Semantic.xsd:
```diff
@@ -659,6 +659,15 @@
 		</xsd:complexContent>
 	</xsd:complexType>	
+
+	<xsd:element name="businessRule" type="tBusinessRule" substitutionGroup="flowElement"/>
```
Этот файл мы обычно не меняем, т.к. это файл схемы-стандарта BPMN
Вижу, что есть `businessRuleTask`, он чем-то не подошел?

zkvan replied:
Отменил изменения, сделал с помощью businessRuleTask, вроде все ок

AlexMisha commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,561 @@
+import ru.runa.gpd.util.VariableUtils;
+import ru.runa.gpd.util.GroovyStuff.Item;
+
+public class BusinessRuleEditorDialog extends Dialog {
```
Насколько понимаю, класс +/- полностью копирует `GroovyEditorDialog`?
Может, получится выделить общую часть, а остальное с помощью наследования/делегирования выразить?

И есть явные несоответствия код стайлу, просьба сверить весь pr с https://runawfe.ru/DevelopersInfo#CodeStyle;

zkvan replied:
кодстайл поправил. На счет наследования, да, могу исправить, но поля классов отличаются, это не проблема?

AlexMisha replied:
Тут скорее надо выделять общую часть и с помощью наследования кастомизировать поведение

AlexMisha commented on plugins/ru.runa.gpd/src/ru/runa/gpd/lang/action/ChooseDelegableHandlerAction.java:
```diff
@@ -4,17 +4,21 @@
+	@Override
+	public void run(IAction action) {
+		Delegable delegable = (Delegable) getSelection();
+		if (!(delegable instanceof BusinessRule)) {
```
По-моему было достаточно поставить `actioHandler='false'` в конфигурациии

zkvan replied:
Это было заглушкой, не мог разобраться как убрать лишний пункт меню, сделал просто не активным. Сейчас исправил

AlexMisha commented on plugins/ru.runa.gpd/plugin.xml:
```diff
@@ -1607,6 +1636,16 @@
+            cellEditorProvider="ru.runa.gpd.extension.businessRule.BusinessRuleProvider"
+            className="ru.runa.wfe.extension.handler.var.BusinessRuleHandler"
+            decisionHandler="false"
+            enabled="false"
```
А точно ли нужно прописывать сюда обработчик, если он не будет использоваться ни из задач сценария, ни из задач ботов?

zkvan replied:
Если сделать actioHandler='false' в DS всплывает замечание - "класс обработчика не найден". А без этой конфигурации вообще не работает %)

Commit:
Отменил исправления semantic.xsd, использовал businessRuleTask, он
подошел.
Удалил лишний пункт меню бизнес-правила "выбрать класс обработчика".
Исправил codeStyle

Commit:
1

Commit:
Поменял иконку

Commit:
внес изменения сервера в jar-ник wfe-core
поменял иконку бизнесс-правила

Commit:
Исправил огибку с пересечением имени переменной и значения

AlexMisha commented on plugins/ru.runa.gpd/plugin.xml:
```diff
@@ -1181,6 +1192,24 @@
+      <element
+            bpmn="businessRuleTask"
+            enabledInRegulationsByDefault="false"
+            icon="bussiness-pravilo.png"
```
`business-rule.png`? И сам файл соответственно тоже желательно переименовать

AlexMisha commented on plugins/ru.runa.gpd/src/ru/runa/gpd/lang/action/ChooseDelegableHandlerAction.java:
```diff
@@ -4,10 +4,11 @@
+import ru.runa.gpd.lang.model.bpmn.BusinessRule;
 import ru.runa.gpd.ui.dialog.ChooseHandlerClassDialog;
 
 public class ChooseDelegableHandlerAction extends BaseModelActionDelegate {
```
Лишние diff'ы в классе

AlexMisha replied:
Все еще лишние diff'ы

zkvan replied:
похоже не запушил, исправлю

zkvan replied:
в проекте удалено, здесь почему-то не отображается...

CrimsonParade replied:
Возможно, автоформатирование в IDE?

zkvan replied:
В форке репо "zin1995 / runawfe-free-devstudio" лишнего нету, не могу понять почему тут отображается.

Commit:
исправил имя файла business-rule.png.
удалил лишний импорт в ChooseHandlerClassDialog

Commit:
далил лишнюю локализацию

Commit:
произвел рефакторинг BusinessRuleEditorDialog и GroovyEditorDialog,
выделил общую часть в GroovyEditorDialogType

AlexMisha commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleModel.java:
```diff
@@ -0,0 +1,248 @@
+import com.google.common.base.Strings;
+import com.google.common.collect.Sets;
+
+public class BusinessRuleModel implements GroovyModel {
```
Класс сильно отличается от `GroovyDecisionModel`? По идее не должен, стоит выделить общую часть

zin1995 replied:
Они вообще практически идентичны, в основном только строка с именем transition заменена на строку function. Делема, как назвать эту переменную в их общем классе?)

zkvan replied:
Этот момент пока отложу, получил задачу, скорее всего буду сильно менять BusinessRuleModel

CrimsonParade replied:
Все еще актуально, в классах достаточно много общего

zkvan replied:
Если честно, не совсем понимаю стоит ли это делать. Классы похожи структурно, но отличаются логически. Мне кажется если их унифицировать, это только усложнит читаемость и понимание кода.

CrimsonParade replied:
Глубоко не вникал, но на первый взгляд не видно каких-то серьезных различий в бизнес-логике классов.
Можешь расписать чуть подробнее, в чем они состоят?

zkvan replied:
Мы можем пообщаться в телеграмме?

zkvan replied:
В основном разный нейминг методов, связанный с тем, что в одном случае мы работаем с "именем перехода", а в другом с формулой. Плюс наличие во внутреннем классе различающийся полей, то есть обращаться к ним придется все равно явно указывая тип класса. Различия в конструкторе плотно связанны с телом, то есть выделить общую часть проблематично, разве что выносить ее в отдельный метод. Мне кажется выделение общей части только усложнит понимание логики обоих классов.

CrimsonParade replied:
Да, можно в телегу: t.me/ExDeathX

Если взять, например, конструктор, то там разница только 
в названии переменной (`function` вместо `transition`) и 
в том, где в объекте хранится эта переменная (инициализация `defaultFunction` вместо `addIfExpr()`).
Имхо, выделить общую часть будет полезно, ибо классы, по сути своей, занимаются одним и тем же, но с разными объектами предметной области => в обсуждаемых классах логичнее видеть только специфические поля/методы, а не дублирование функционала.

zkvan replied:
Конструктор GroovyDecidionModel не поддерживает множественные условия, соответственно механизм парсинга отличается. Из всего класса, безболезненно, получится выделить в общий класс только get и setIfExp методы, и normalizeString. Остальное завязано на различие во внутреннем классе (тоже используются массивы для переменных, вместо их одиночного хранения) и на разнице в нейминге (function и transition), которые тоже на мой взгляд менять нельзя, тк имена говорящие.

AlexMisha commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,310 @@
+        createButtomComposite();
+    }
+
+    private Composite createButtomComposite() {
```
```suggestion
    private Composite createBottomComposite() {
```

zkvan replied:
спасибо! я пол часа убил чтобы понять в чем разница в строках, не представляю как ты смог найти эту опечатку в коде)

AlexMisha commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,310 @@
+    }
+
+    private void createLines(String function) {
+        FilterBox[] var = new FilterBox[2];
```
`var` - не слишком удачное имя, стоит отразить в названии, зачем нужна переменная

AlexMisha commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,310 @@
+        var[0].setSelectionListener(new FilterBoxSelectionHandler());
+        var[0].setLayoutData(getGridData());
+
+        Combo oper = new Combo(constructor, SWT.READ_ONLY);
```
`operations`? Сокращения иногда могут сбить с толку

AlexMisha commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,310 @@
+            IfExpr ifExpr = ((BusinessRuleModel) initModel).getIfExpr(functions.get(i));
+            if (ifExpr != null) {
+                buttons.get(i).setText(ifExpr.getFunction());
+                Variable variable = ifExpr.getVariable1();
```
Тоже просьба отразить в названии, что это за переменная и что за индекс идет ниже

zkvan replied:
Тут просто локальная переменная для заполнения FilterBox, index тоже служебная переменная в FilterBox - просто порядковый номер элемента в массиве. То есть более говорящих имен и не подберешь.

AlexMisha commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,310 @@
+    protected void refresh(FilterBox filterBox) {
+        try {
+            int[] indexes = (int[]) filterBox.getData(DATA_INDEX_KEY);
+            if (indexes[1] == 2) {
```
`indexes` - название стоит конкретизировать
И почему == именно 2? Мне это кажется не очевидным, стоит оставить комментарий

zkvan replied:
тоже просто локальная переменная индекса поля, не знаю даже нужно ли конкретизировать

CrimsonParade replied:
Может быть `variableIndexes`?

Commit:
исправил имена переменных на более говорящие, исправил опечатку в имени
createBottomComposite()

Commit:
небольшой рефакторинг, сделал решение без лишнего листа (fancrions)

Commit:
Графический интерфейс для сложных условий

Commit:
1

Commit:
Реализована возможность добавления сложных условий в элементе
бизнес-правило

Commit:
исправление выявленных ошибок

Commit:
удалил лишние коментарии

Commit:
Исправил работу с Date и BigDecimal переменными

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,534 @@
+        @Override
+        public void widgetSelected(SelectionEvent e) {
+            FilterBox fb = ((FilterBox) e.widget);
+            FilterBox targetFB;
```
Аббревиатуры тоже следует писать в camel case: "targetFB -> targetFb"
https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

Тут, на мой взгляд, лучше будет вообще без оных: `fb`/`targetFB` -> `box`/`targetBox`

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,534 @@
+            FilterBox fb = ((FilterBox) e.widget);
+            FilterBox targetFB;
+            int[] indexData = (int[]) fb.getData(DATA_INDEX_KEY);
+            int varNum = indexData[2] == 0 ? 0 : 1;
```
Множество `Magic Numbers`, которые используются по всему классу, может быть стоит вынести в константы с понятными названиями? (тогда и комментировать их не нужно будет)
Читаться будет в разы быстрее и проще

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,534 @@
+        }
+    }
+
+    public class SyncOperComboHandler extends SelectionAdapter {
```
Желательно избегать сокращений, особенно в названиях классов/методов

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,534 @@
+        }
+    }
+
+    private void createExpr(Composite parent, int indexLine) {
```
`Expression` тоже лучше писать полностью (и в названии метода, и в названиях переменных)

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,534 @@
+
+        FilterBox[] variable = new FilterBox[2];
+        exprLine.setVariableBox(variable);
+        int indexExpr = exprLine.getVariableBoxes().indexOf(variable);
```
`expressionIndex` будет правильнее

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,534 @@
+        }
+        exprComposite.setLayoutData(data);
+
+        FilterBox[] variable = new FilterBox[2];
```
`variables`?

CrimsonParade replied:
Как-то неправильно понял или все-таки это массив переменных?

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,534 @@
+
+        Combo logicBox = new Combo(exprComposite, SWT.READ_ONLY);
+        exprLine.setLogicBox(logicBox);
+        logicBox.setItems("null", "and", "or");
```
Как и с `Magic Numbers`, следует вынести в константы

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,534 @@
+        try {
+            int[] indexes = (int[]) filterBox.getData(DATA_INDEX_KEY);
+            if (indexes[2] == 2) {
+                // indexes[1] == 2 - second variable
```
Последствия `Magic Numbers`: не только приходится писать комментарии, но они еще и расходятся с кодом
 (в коде - `indexes[2]`, в комменте - `indexes[1]`)

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,534 @@
+        for (int i = 0; i < exprLines.size(); i++) {
+            ExprLine exprLine = exprLines.get(i);
+            for (int j = 0; j < exprLine.getVariableBoxes().size(); j++) {
+                if (exprLine.getVariableBoxes().get(j)[0].getText().length() == 0 || exprLine.getVariableBoxes().get(j)[1].getText().length() == 0
```
Достаточно длинное условие, заставляющее долго в него вникать и писать комментарии. 
На мой взгляд, лучше вынести в переменную (что-то вроде `dataIsEmpty`) - тогда и комментарий будет излишним, и читать будет проще.

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,534 @@
+            for (int i = 0; i < exprLines.size(); i++) {
+                ExprLine exprLine = exprLines.get(i);
+                IfExpr ifExpr;
+                List<Variable> variable1list = new ArrayList();
```
Не очень нравится название, может быть что-то вроде `firstVariables` или `leftVariables`?

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,534 @@
+                ExprLine exprLine = exprLines.get(i);
+                IfExpr ifExpr;
+                List<Variable> variable1list = new ArrayList();
+                List<Object> lexem2list = new ArrayList();
```
`secondLexems`/`rightLexems`?

CrimsonParade reviewed (changes requested):
Каждое сокращение отмечать не стал, просьба пройтись по всему классу и попереименовывать в полные формы (особенно `Expr` -> `Expression`)

Commit:
привел код к более читаемому виду

Commit:
1

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,542 @@
+        }
+    }
+
+    private static final int FIRST_VARIABLE_INDEX = 0;
```
Константы принято располагать в начале класса (над "обычными" полями)

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,542 @@
+                    expressionLine.getVariableBoxes().get(j)[0].select(firstVariableIndex);
+                    refresh(expressionLine.getVariableBoxes().get(j)[0]);
+                    GroovyTypeSupport typeSupport = GroovyTypeSupport.get(firstVariable.getJavaClassName());
+                    firstVariableIndex = Operation.getAll(typeSupport).indexOf(ifExpr.getOperations().get(j));
```
Думаю, не стоит переиспользовать `firstVariableIndex`, лучше будет добавить `operationIndex`.
В противном случае, если читать по диагонали, далеко не очевидно, почему условие `(firstVariableIndex == -1)` встречается дважды.

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,542 @@
+                    refresh(expressionLine.getOperationBoxes().get(j));
+                    String secondVariableText = ifExpr.getSecondVariableTextValue(j);
+                    int secondVariableIndex = 0;
+                    if (VariableUtils.getVariableByScriptingName(variables, secondVariableText) != null
```
Достаточно объемное условие - стоит вынести в переменную (по аналогии с `emptyFieldExist`)

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,542 @@
+                return;
+            }
+            if (indexes[2] == FIRST_VARIABLE_INDEX) {
+                Combo operCombo = expressionLines.get(indexes[0]).getOperationBoxes().get(indexes[1]);
```
`operationCombo`/`operations`

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/businessRule/BusinessRuleEditorDialog.java:
```diff
@@ -0,0 +1,542 @@
+                    String operationName = expressionLine.getOperationBoxes().get(j)
+                            .getItem(expressionLine.getOperationBoxes().get(j).getSelectionIndex());
+                    String secondVariableText = expressionLine.getVariableBoxes().get(j)[1].getText();
+                    Object secondVariable;
```
`Object` как тип переменной не нравится, да и название `variable2` возникает из-за этого же.

Кажется, так будет лучше (вместо вязанки if'ов, связанной с инициализацией `secondVariable`):
```suggestion
                    Variable secondVariable = VariableUtils.getVariableByScriptingName(variables, secondVariableText);  
                    if (secondVariable != null && secondVariable.getJavaClassName().equals(firstVariable.getJavaClassName())) {
                        secondVariables.add(secondVariable);
                    } else {
                        secondVariables.add(secondVariableText);
                    }
```

Ниже, соответсвенно, `secondVariables.add(secondVariable);` убрать

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/decision/GroovyEditorDialogType.java:
```diff
@@ -0,0 +1,289 @@
+    protected void refresh(FilterBox filterBox) {
+    };
+
+    protected List<String> getSecondVariableNames(Variable variable1) {
```
Обычно, если в названии появляются числа, можно придумать что-то более подходящее.

Тут, судя по использованию, в качестве параметра приходит `firstVariable`?
Если так, то напрашивается переименование `variable1` -> `firstVariable`, 
а ниже, соответственно, `typeSupport1` -> `firstVariableTypeSupport`.

zkvan replied:
Подобный нейминг выбирал не я, просто перенес его при выделении общей части классов. Стоит ли мне тогда внести аналогичные изменения в GroovyEditorDilog и в GroovyDecisionModel?

CrimsonParade replied:
Да, к сожалению, в старом коде тоже встречаются проблемы :)
Общее правило звучит примерно так: 
"Если проблемы находятся в той части, что изменяется, попутный рефакторинг - добро, если нет - трогать не стоит (лишние мерж-конфликты)".
`GroovyEditorDialog`, судя по diff'y, постигли довольно серьезные изменения - думаю, самые простые ренеймы не будут лишними.
 `GroovyDecisionModel`, если решим выделять общую часть, тоже весьма сильно изменится => при выделении общей части можно будет позволить себе подпричесать и код в целом, и названия в частности.

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/decision/GroovyEditorDialogType.java:
```diff
@@ -0,0 +1,289 @@
+    protected void initConstructorView() {
+    };
+
+    protected static final String DATA_INDEX_KEY = "indexes";
```
Аналогично `BusinessRuleEditorDialog`, константы стоит перенести наверх.

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/decision/GroovyEditorDialogType.java:
```diff
@@ -0,0 +1,289 @@
+                }
+            }
+        } catch (RuntimeException e) {
+            // Activate source view if custom code found
```
Выглядит как `Control Flow Exceptions` - стоит подумать, как можно от этого избавиться.

zkvan replied:
В данном случае это выглядит каким-то костылем, без какой-либо полезного применения. Решил просто удалить)

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/decision/GroovyEditorDialogType.java:
```diff
@@ -0,0 +1,289 @@
+        styledText.addLineStyleListener(new JavaHighlightTextStyling(variableNames));
+        styledText.setText(this.initValue);
+        styledText.setLayoutData(new GridData(GridData.FILL_BOTH));
+        TabItem tabItem2 = new TabItem(tabFolder, SWT.NONE);
```
Более конкретное название? (можно отталкиваться от `GroovyEditor.title.code`)

CrimsonParade commented on plugins/ru.runa.gpd/src/ru/runa/gpd/extension/decision/GroovyEditorDialogType.java:
```diff
@@ -0,0 +1,289 @@
+        tabFolder.addSelectionListener(new TabSelectionHandler());
+        Composite constructorView = new Composite(tabFolder, SWT.NONE);
+        constructorView.setLayout(new GridLayout());
+        TabItem tabItem1 = new TabItem(tabFolder, SWT.NONE);
```
Более конкретное название? (можно отталкиваться от `GroovyEditor.title.constructor`)

CrimsonParade reviewed (changes requested):
Плюс осталось два нерешенных вопроса с предыдущего ревью: 
если пропущено сознательно, просьба написать, по какой причине.

Commit:
Рефакторинг/ренейминг

Commit:
Рефакторинг\ренейминг

Commit:
1

Commit:
Рефакторинг\ренейминг

Commit:
Рефакторинг BusinessRuleEditorDialog

Commit:
Ренейминг полей GroovyDecisionModel и GroovyEditorDialog

Commit:
1

Commit:
1

Commit:
1

Commit:
1

Commit:
Рефакторинг/ренейминг

Commit:
Рефакторинг/ренейминг

Commit:
Рефакторинг

Commit:
test

Commit:
end test
